build/patches/Timezone-Override.patch

```diff
@@ -0,0 +1,1379 @@
++++ b/components/content_settings/renderer/content_settings_agent_impl.cc
+@@ -38,6 +38,11 @@
+ #include "url/gurl.h"
+ #include "url/origin.h"
+ #include "url/url_constants.h"
++#include "third_party/blink/renderer/core/timezone/timezone_controller.h"
++#include "third_party/blink/renderer/core/inspector/locale_controller.h"
++#include "third_party/blink/renderer/core/inspector/locale_controller.h"
++#include "third_party/icu/source/common/unicode/strenum.h"
++#include "third_party/icu/source/i18n/unicode/timezone.h"
+ 
+ using blink::WebDocument;
+ using blink::WebFrame;
+@@ -48,6 +53,8 @@ using blink::WebURL;
+ using blink::WebView;
+ using content::DocumentState;
+ 
++std::unique_ptr<blink::TimeZoneController::TimeZoneOverride> timezone_override_;
++
+ namespace content_settings {
+ namespace {
+ 
+@@ -346,6 +353,9 @@ bool ContentSettingsAgentImpl::AllowScript(bool enabled_per_settings) {
+   allow = allow || IsWhitelistedForContentSettings();
+ 
+   cached_script_permissions_[frame] = allow;
++  
++  if (allow) UpdateOverrides();
++
+   return allow;
+ }
+ 
+@@ -572,4 +582,104 @@ bool ContentSettingsAgentImpl::AllowStorageAccess(
+   return result;
+ }
+ 
++bool ContentSettingsAgentImpl::UpdateOverrides() {
++  // Evaluate the content setting rules
++  ContentSetting setting = CONTENT_SETTING_ALLOW;
++
++  if (content_setting_rules_) {
++    blink::WebLocalFrame* frame = render_frame()->GetWebFrame();
++
++    setting = GetContentSettingFromRules(
++        content_setting_rules_->timezone_override_rules, frame,
++        url::Origin(frame->GetDocument().GetSecurityOrigin()).GetURL());
++  }
++  return UpdateTimeZoneOverride(setting, content_setting_rules_->timezone_override_value);
++         //&& UpdateLocaleOverride(setting);
++}
++
++// bromite #331: Customize Timezone
++bool ContentSettingsAgentImpl::UpdateTimeZoneOverride(ContentSetting setting, const std::string& timezone_override_value) {
++  // base/i18n/icu_util.cc # 329
++  //VLOG(0) << "---Bromite:ContentSettingsAgentImpl::UpdateTimeZoneOverride start";
++  //VLOG(0) << "---" << timezone_override_value;
++
++  /* timezone_id: third_party/icu/source/i18n/timezone.cpp
++      We first try to lookup the zone ID in our system list.  If this
++      * fails, we try to parse it as a custom string GMT[+-]hh:mm.  If
++      * all else fails, we return GMT, which is probably not what the
++      * user wants, but at least is a functioning TimeZone object.
++      */
++  String timezone_id;
++
++  if(setting == CONTENT_SETTING_ALLOW) {
++    // system time
++    if(timezone_override_) {
++      timezone_override_.reset();
++    }
++    return true;
++  } else if(setting == CONTENT_SETTING_BLOCK) {
++    // timezone random
++    UErrorCode ec = U_ZERO_ERROR;
++    int32_t rawOffset = ((rand() % 24)-12) * 3600 * 1000;
++    icu::StringEnumeration* timezones = icu::TimeZone::createEnumeration(rawOffset); // Obtain timezones by GMT timezone offset
```

Comment:
Instead of these we should pick from a fixed list of timezones.

I have created a shortlist of most common ones:
```
  'Asia/Dubai',
  'Europe/Tirane',
  'America/Argentina/Buenos_Aires',
  'Europe/Vienna',
  'Australia/Melbourne',
  'Australia/Sydney',
  'Europe/Brussels',
  'Europe/Sofia',
  'America/Sao_Paulo',
  'America/Toronto',
  'America/Vancouver',
  'Europe/Zurich',
  'America/Santiago',
  'Asia/Shanghai',
  'America/Bogota',
  'Europe/Prague',
  'Europe/Berlin',
  'Europe/Copenhagen',
  'Africa/Cairo',
  'Europe/Madrid',
  'Europe/Helsinki',
  'Europe/Paris',
  'Europe/London',
  'Europe/Athens',
  'Asia/Hong_Kong',
  'Europe/Budapest',
  'Asia/Jakarta',
  'Europe/Dublin',
  'Europe/Rome',
  'Asia/Tokyo',
  'Asia/Seoul',
  'America/Mexico_City',
  'Europe/Oslo',
  'America/Lima',
  'Europe/Warsaw',
  'Europe/Lisbon',
  'Europe/Bucharest',
  'Europe/Belgrade',
  'Europe/Moscow',
  'Asia/Singapore',
  'America/El_Salvador',
  'Asia/Bangkok',
  'Europe/Istanbul',
  'America/New_York',
  'America/Detroit',
  'America/Chicago',
  'America/Denver',
  'America/Phoenix',
  'America/Los_Angeles',
  'Africa/Johannesburg'
```

Reply:
I disagree, why have one more list?
it is already present in the icu library, managed and maintained by them

Reply:
A list of most common timezones; you could also use the full list from the library but it would not be limited to the most common ones.

Reply:
I think it already is.
the library for a timezone with a fixed integer time divisible by 3600 * 1000 (to understand therefore +/- 1, +/- 2 etc.) gives a list of available timezones, I always take the first one.

Reply:
I see; that should work.